components/sup/src/manager/mod.rs

```diff
@@ -104,74 +115,73 @@ impl Manager {
             // Note - eventually, we need to deal with suitability here. The original implementation
             // didn't have this working either.
-            self.butterfly.start_election(service_group.clone(), 0, 0);
+            self.state.butterfly.start_election(service_group.clone(), 0, 0);
         }
 
         let service = try!(Service::new(service_group, package, topology, update_strategy));
-        self.services.push(service);
-
+        self.updater.add(&service);
+        self.state.services.write().unwrap().push(service);
         Ok(())
     }
 
     pub fn build_census(&mut self, last_update: &CensusUpdate) -> Result<(bool, CensusUpdate)> {
-        let update = CensusUpdate::new(self.butterfly.service_store.get_update_counter(),
-                                       self.butterfly.election_store.get_update_counter(),
-                                       self.butterfly.member_list.get_update_counter());
+        let update = CensusUpdate::new(self.state.butterfly.service_store.get_update_counter(),
+                                       self.state.butterfly.election_store.get_update_counter(),
+                                       self.state.butterfly.member_list.get_update_counter());
 
         if &update != last_update {
             let mut cl = CensusList::new();
             debug!("Updating census from butterfly data");
-            self.butterfly.service_store.with_keys(|(service_group, rumors)| {
+            self.state.butterfly.service_store.with_keys(|(service_group, rumors)| {
                 for (member_id, service) in rumors.iter() {
                     let mut ce = CensusEntry::default();
                     ce.populate_from_service(service);
-                    cl.insert(String::from(self.butterfly.member_id()), ce);
+                    cl.insert(String::from(self.state.butterfly.member_id()), ce);
                 }
             });
-            self.butterfly.election_store.with_keys(|(service_group, rumors)| {
+            self.state.butterfly.election_store.with_keys(|(service_group, rumors)| {
                 // We know you have an election, and this is the only key in the hash
                 let election = rumors.get("election").unwrap();
                 cl.populate_from_election(election);
             });
-            self.butterfly.member_list.with_members(|member| {
+            self.state.butterfly.member_list.with_members(|member| {
                 cl.populate_from_member(member);
             });
-            self.census_list = cl;
+            *self.state.census_list.write().unwrap() = cl;
             return Ok((true, update));
         }
-
         Ok((false, update))
     }
 
     fn check_for_incoming_signals(&mut self) -> bool {
         match signals::check_for_signal() {
             Some(SignalEvent::Shutdown) => {
-                for service in self.services.iter_mut() {
+                for service in self.state.services.write().unwrap().iter_mut() {
```

Comment:
These unwrap's all need to become `expect("Services lock is poisoned")` or something similar, rather than just the raw explosion. Makes a big difference in debug hunting - the raw unwrap tells you almost nothing about the lock you're digging for.